File: bst.py
class Node:
    def __init__(self, val):
        self.value = val
        self.leftChild = None
        self.rightChild = None

    def insert(self, data):
        if (self.value == data):
            return False

        elif self.value > data:
            if self.leftChild:
                return self.leftChild.insert(data)
            else:
                self.leftChild = Node(data)
                return True

        elif self.value < data:
            if self.rightChild:
                return self.rightChild.insert(data)
            else:
                self.rightChild = Node(data)
                return True

    def find(self, data):
        if (self.value == data):
            return True

        elif self.value > data:
            if self.leftChild:
                self.leftChild.find(data)
            else:
                return False

        elif self.value < data:
            if self.rightChild:
                self.rightChild.find(data)
            else:
                return False

    def getHeight(self):
        if self.leftChild and self.rightChild:
            return 1 + max(self.leftChild.getHeight(), self.rightChild.getHeight())
        elif self.leftChild:
            return 1 + self.leftChild.getHeight()
        elif self.rightChild:
            return 1 + self.rightChild.getHeight()
        else:
            return 1

    def getSize(self):
        if self.leftChild and self.rightChild:
            return 1 + self.leftChild.getSize() + self.rightChild.getSize()
        elif self.leftChild:
            return 1 + self.leftChild.getSize()
        elif self.rightChild:
            return 1 + self.rightChild.getSize()
        else:
            return 1

    def preorder(self):
        if self:
            print (str(self.value))
            if self.leftChild:
                self.leftChild.preorder()
            if self.rightChild:
                self.rightChild.preorder()

    def postorder(self):
        if self:
            if self.leftChild:
                self.leftChild.postorder()
            if self.rightChild:
                self.rightChild.postorder()
            print (str(self.value))

    def inorder(self):
        if self:
            if self.leftChild:
                self.leftChild.inorder()
            print (str(self.value))
            if self.rightChild:
                self.rightChild.inorder()

class Tree:
    def __init__(self):
        self.root = None

    def insert(self, data):
        if self.root:
            return self.root.insert(data)
        else:
            self.root = Node(data)
            return True

    def find(self, data):
        if self.root:
            return self.root.find(data)
        else:
            return False

    def getHeight(self):
        if self.root:
            return self.root.getHeight()
        else:
            return -1

    def getSize(self):
        if self.root:
            return self.root.getSize()
        else:
            return 0

    def remove(self, data):
        if self.root is None:
            return False

        elif self.root.value == data:

            if self.root.leftChild is None and self.root.rightChild is None:
                self.root = None

            elif self.root.leftChild and self.root.rightChild is None:
                self.root = self.leftChild

            elif self.root.leftChild is None and self.root.rightChild:
                self.root = self.rightChild

            elif self.root.leftChild and self.root.rightChild:
                delNodeParent = self.root
                delNode = self.root.rightChild
                while delNode.leftChild:
                    delNodeParent = delNode
                    delNode = delNodeParent.leftChild

                self.root.value = delNode.value
                if delNode.rightChild:
                    if delNodeParent.value > delNode.value:
                        delNodeParent.leftChild = delNode.rightChild

                    elif delNodeParent.value < delNode.value:
                        delNodeParent.rightChild = delNode.rightChild

                    else:
                        if delNodeParent.value > delNode.value:
                            delNodeParent.leftChild = None
                        elif delNodeParent.value < delNode.value:
                            delNodeParent.rightChild = None

                return True

        parent = None
        node = self.root

        # find node to remove
        while node and node.value != data:
            parent = node
            if data < node.value:
                node = node.leftChild
            if data > node.value:
                node = node.rightChild

        # Case 1: data not found
        if node is None or node.value != data:
            return False

        # Case 2: remove-node has no children
        elif node.leftChild is None and node.rightChild is None:
            if data < parent.value:
                parent.leftChild = None
            if data > parent.value:
                parent.rightChild = None
            return True

        # Case 3: remove-node has left child only
        elif node.leftChild and node.rightChild is None:
            if data < parent.value:
                parent.leftChild = node.leftChild
            if data > parent.value:
                parent.rightChild = node.leftChild
            return True

        # Case 4: remove-node has right child only
        elif node.rightChild and node.leftChild is None:
            if data < parent.value:
                parent.leftChild = node.rightChild
            if data > parent.value:
                parent.rightChild = node.rightChild
            return True

        # Case 5: remove-node has both left and right children
        # You replace remove-node with either the left most child on its right side,
        # or the right most child on its left side.
        # You then delete the child from the bottom that it was replaced with.

        else:
            delNodeParent = node
            delNode = node.rightChild
            while delNode.leftChild:
                delNodeParent = delNode
                delNode = delNode.leftChild

            node.value = delNode.value
            if delNode.rightChild:
                if delNodeParent.value > delNode.value:
                    delNodeParent.leftChild = delNode.rightChild
                elif delNodeParent.value < delNode.value:
                    delNodeParent.rightChild = delNode.rightChild
            else:
                if delNodeParent.value > delNode.value:
                    delNodeParent.leftChild = None
                else:
                    delNodeParent.rightChild = None

    def preorder(self):
        print ("PreOrder")
        self.root.preorder()

    def postorder(self):
        print ("PostOrder")
        self.root.postorder()

    def inorder(self):
        print ("InOrder")
        self.root.inorder()

bst = Tree()
print (bst.insert(8))
bst.insert(6)
bst.insert(10)
bst.insert(7)
bst.insert(5)
bst.insert(9)
bst.insert(11)
bst.remove(11)

bst.preorder()
bst.postorder()
bst.inorder()
print(bst.getHeight())
print('Size = ', bst.getSize())
# Output
# True
# PreOrder
# 8
# 6
# 5
# 7
# 10
# 9
# 11
# PostOrder
# 5
# 7
# 6
# 9
# 11
# 10
# 8
# InOrder
# 5
# 6
# 7
# 8
# 9
# 10
# 11
# 3

File: bst.cpp
#include <iostream>
#include <queue>
using namespace std;

struct Node {
  int data;
  Node *left, *right;
};

Node* GetNewNode(int data) {
  Node *newNode = new Node();
  newNode->data = data;
  newNode->left = newNode->right = NULL;
  return newNode;
}

Node* Insert(Node *root, int data){
  if(root == NULL){
    root = GetNewNode(data);
  }
  else if(data <= root->data){
    root->left = Insert(root->left, data);
  }
  else{
    root->right = Insert(root->right, data);
  }
  return root;
}

bool Search(Node *root, int data){
  if (root == NULL){
    return false;
  }
  else if (root->data == data){
    return true;
  }
  else if (data <= root->data){
    return Search(root->left, data);
  }
  else{
    return Search(root->right, data);
  }
}

Node* findMin(Node *root){
  if(root == NULL) return;
  Node* current = root;
  while(current->left != NULL){
    current = current->left;
  }
  return current;
}

Node* findMax(Node *root){
  // Using iteration
  if(root == NULL) return;
  Node* current = root;
  while(current->right != NULL){
    current = current->right;
  }
  return current;
}

Node* findMax2(Node *root){
  // Using recursion
  if(root == NULL) return;
  else if(root->right == NULL){
    return root;
  }
  return findMax(root->right);
}

Node* Delete(Node *root, int data){
  if(root == NULL) return root;
  else if(data < root->data){
    root->left = Delete(root->left, data);
  }
  else if(data > root->data){
    root->right = Delete(root->right, data);
  }
  // When node to be deleted is found
  else{
    // Case1: No child
    if(root->left == NULL && root->right == NULL){
      delete root;
      root = NULL;
    }
    // Case2: One child
    else if(root->left == NULL){
      Node *temp = root;
      root = root->right;
      delete temp;
    }
    else if (root->right == NULL){
      Node *temp = root;
      root = root->left;
      delete temp;
    }
    // Case3: 2 children
    else{
      Node *temp = findMin(root->right);
      // Node *temp = findMax(root->left);
      root->data = temp->data;
      root->right = Delete(root->right, temp->data);
    }
  }
  return root;
}

void InOrder(Node *root){
  if(root==NULL) return;
  InOrder(root->left);
  cout << root->data << " ";
  InOrder(root->right);
}

bool isIdentical(Node *root1, Node *root2){
  if(root1 == NULL && root2 == NULL){
    return true;
  }
  else if((root1->data == root2->data) &&
          isIdentical(root1->left, root2->left) &&
          isIdentical(root1->right, root2->right)){
            return true;
  }
  else{
    return false;
  }
}

int findHeight(Node *root){
  // Returns no. of edges in longest path from root to a leaf node.
  if(root==NULL) return -1;
  return max(findHeight(root->left),findHeight(root->right))+1;
}

void LevelOrder(Node* root){
  if(root == NULL) return;
  queue<Node*> Q;
  Q.push(root);
  while(!Q.empty()){
    Node* current = Q.front();
    cout << current->data << endl;
    if(current->left != NULL) Q.push(current->left);
    if(current->right != NULL) Q.push(current->right);
    Q.pop();
  }
}

bool isLeaf(Node* root){
  return (root->left == NULL and root->right == NULL);
}

bool isBST(Node *root, Node *l = NULL, Node *r = NULL){
    if (root == NULL) return true;
    if (l != NULL and root->data < l->data) return false;
    if (r != NULL and root->data > r->data) return false;
    return isBST(root->left, l, root) and isBST(root->right, root, r);
}

bool isIdentical(Node *root1, Node* root2){
  if(root1 != root2){
    return false;
  }
  else if((root1->data == root2->data) &&
           isIdentical(root1->left, root2->left) &&
           isIdentical(root1->right, root2->right)){
             return true;
  }

int main() {

  Node *root = NULL;
  root = Insert(root,15);
	root = Insert(root,10);
	root = Insert(root,20);
	root = Insert(root,25);
	root = Insert(root,8);
	root = Insert(root,12);

  Node *root1 = NULL;
  root1 = Insert(root1,15);
	root1 = Insert(root1,10);
	root1 = Insert(root1,20);
	root1 = Insert(root1,25);
	root1 = Insert(root1,8);
	root1 = Insert(root1,12);

  if (isIdentical(root, root1)){
    cout << "Identical";
  }
  else{
    cout << "Not identical";
  }

  // root = Delete(root,25);
  //
  // int number;
  // cout << "Enter a number to search\n";
  // cin >> number;
  //
  // if (Search(root, number)==true){
  //   cout << "Found\n";
  // }
  // else{
  //   cout << "Not Found\n";
  // }
  // cout << "Min : " << findMin(root)->data << endl;
  // cout << "Max : " << findMax(root)->data << endl;
  //
  // if (isBST(root)){
  //   cout << "Is BST" << endl;
  // }
  // else{
  //   cout << "Not a BST" << endl;
  // }
  //
  // cout << "Height of BST : " << findHeight(root) << endl;
  //
  // cout << "Inorder: ";
  // InOrder(root);
  // cout << "\n";
}
